### ISSUE TITLE
custom mouse functions

### ISSUE THREAD

**Issue by kne42** [napari/napari#252]
as discussed in the meeting today, we may want to support user-defined functions that get executed with mouse events

following is the proposed format of the function a user would provide:
```python
def status(layer):
    # on mouse press
    print('pressed!')

    yield

    while layer.dragging:
        # on mouse move
        print(layer.pos)
        yield

    # on mouse release
    print('released!')
```

**Comment by sofroniewn** [comment]
Looking promising!
 
We need to be able to get both an `is_moving` and an `is_dragging` (which is a subset of `is_moving`)

And we need to get any modifier keys (like if `Shift` being pressed)

Also we should make sure that we are ok not supporting `double_click` events for now. One of the reasons is that `double_click` events can be handled different on different OS - read this disclaimer from the vispy docs

> The ordering of the mouse_double_click, mouse_press, and mouse_release events are not guaranteed to be consistent between backends. Only certain backends natively support double-clicking (currently Qt and WX); on other backends, they are detected manually with a fixed time delay. This can cause problems with accessibility, as increasing the OS detection time or using a dedicated double-click button will not be respected.

Any further thoughts from @royerloic @jni?

**Comment by sofroniewn** [comment]
Another idea was to have an extra mode / button on the left for each layer that allows the "custom cursor" to be selected. When in this mode any custom mouse events get executed if they have been defined. What this means is that the normal cursor is always the normal cursor with pan/zoom, the paintbrush is always the paintbrush, and the "custom" is always user defined.

If the user wants multiple user defined cursors, they can all live under the custom cursor mode / button and they can then use custom hotkeys to switch between them.

@gregjohnso does this sound like useful functionality to complement the custom keybindings?

**Comment by royerloic** [comment]
I think we can definitely live without double clicking, at least for now.
It might even be a good thing, double clicking is not a necessity and
can be unnecessarily strenuous if needed many times -- think
annotations, etc...

**Comment by jni** [comment]
Well, as I mentioned originally, I think double click to zoom is very nice and matches what Google Maps does. From your documentation link, @sofroniewn, it sounds like double click *would* work on Qt?

I don't love the custom cursor button... I'd rather a warning when your settings shadow existing settings. That's what every app I've seen does. There should also be a fast way to resolve these conflicts one way or the other.

When it comes to much fancier modes, they should be implemented with plugins, which define their own buttons. So this argues that we need to allow plugins to add buttons to the toolbar, which I think is probably a good thing. This PR provides nice machinery to help plugins do this. =)

**Comment by sofroniewn** [comment]
@jni - indeed - i think double click would work - we may just need to take some extra care. we use doubleclicks to expand the layer properties widgets right now (though I think we might want to switch that to a button as it doesn't seem to be very intuitive)

I'm also not wedded to the "custom cursor" button, but it would nice if there was a relatively straight forward way to enable / disable the custom mouse functions (say using a custom keybindings), and to make sure that they didn't interfere with the mouse functions that are currently defined on each layer (when that would not be desired) without resorting to the full plugin machinery. We can revisit though after the basics have been implemented.

**Comment by constantinpape** [comment]
It would be very helpful to specify the behavior of the mouse wheel as well.
For example, I would like to map scrolling through an axis, in particular the z-axis for 3d data, to the `MouseWheel` (and remap zoom to `Shift` + `MouseWheel) or map scrolling to `Shift` + `MouseWheel`.

**Comment by sofroniewn** [comment]
This is closed by #544
